### PR TITLE
Prepared System.Buffer.Primitives for API Review

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
- Unified Positions
- Moved ReadCursorOperations around. These APIs are not general purpose Position APIs. They only work with ReadOnlyBuffer.
- Renamed ReadOnlyBufferReader to BufferReader
- Simplified namespaces

@pakrym, @davidfowl, @joshfree, @terrajobst

### CODE REVIEW

File: src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs [hunk 1]
```diff
@@ -56,7 +56,7 @@ public static int CopyTo(this IMemoryList<byte> list, Span<byte> destination)
             {
                 var current = list.Memory.Span;
                 var index = current.IndexOf(value);
-                if (index != -1) return Collections.Sequences.Position.Create(list, index);
+                if (index != -1) return new Collections.Sequences.Position(list, index);
```
Comment (benaadams) [inline src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs]: `System.Collections.Sequences` is imported, is there a name clash? 

i.e. can this be just `Position`?

Comment (KrzysztofCwalina) [reply src/System.Buffers.Experimental/System/Buffers/BufferExtensions.cs]: Ah, forgot about it. I will remove.

File: tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs [hunk 2]
```diff
@@ -180,5 +181,129 @@ public void Create_WorksWithOwnedMemory()
             b => b.Slice(0, 70).Slice(0, b.End),
             b => b.Slice(70, b.Start)
         };
+
+        static bool TryGetBuffer(Position begin, Position end, out ReadOnlyMemory<byte> data, out Position next)
```
Comment (pakrym) [inline tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs]: Why is this here?

File: tests/System.IO.Pipelines.Tests/TestBufferFactory.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO.Pipelines.Testing;
+using System.Linq;
+using System.Text;
+
+namespace System.IO.Pipelines.Tests
```
Comment (pakrym) [inline tests/System.IO.Pipelines.Tests/TestBufferFactory.cs]: This is useful not on only for pipelines, why is it moved back?

File: src/System.Buffers.Primitives/System/Collections/Position.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            switch (Segment)
+            {
+                case null:
+                    return default;
+                case T segment:
+                    return segment;
+            }
+
+            //ThrowHelper.ThrowInvalidOperationException(ExceptionResource.UnexpectedSegmentType);
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Collections/Position.cs]: Why remove throwing?

File: src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 namespace System.IO.Pipelines
 {
-    public static class ReadCursorOperations
+    public static class PositionOperations
```
Comment (pakrym) [inline src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Why is it moved back to pipelines?

File: tests/System.IO.Pipelines.Tests/TestBufferFactory.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply tests/System.IO.Pipelines.Tests/TestBufferFactory.cs]: It's used just by pipelines tests

File: src/System.Buffers.Primitives/System/Collections/Position.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Collections/Position.cs]: I have no idea. Good catch.

File: tests/System.IO.Pipelines.Tests/TestBufferFactory.cs (hunk 3 above)
Comment (pakrym) [reply tests/System.IO.Pipelines.Tests/TestBufferFactory.cs]: No, see https://github.com/dotnet/corefxlab/blob/master/tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs#L21

Comment (pakrym) [reply tests/System.IO.Pipelines.Tests/TestBufferFactory.cs]: Not sure why it builds but it's used by ROB and RBR tests that are in primitives.

File: src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Well, it's pipelines specific. It does not work with an arbitrary position. It only works with positions of IMemoryList, and possibly other things that ReadOnlyBuffer supports.

Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: It works with everything that `ReadOnlyBuffer` supports that's why it's in buffers. There is no need to have another copy in Pipelines.

File: tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs]: I can move the tests that call Seek to Pipelines.Tests and then eliminate it. What do you think?

File: src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Position is more general than Buffers. Look in System.Collections.Sequences. There is a Hashtable and ArrayList implementation that uses it.

Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Then move these methods to be on ROB

File: tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs (hunk 2 above)
Comment (pakrym) [reply tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs]: These tests were testing internal implementations of `ROB` why would we want to move them to `Pipelines`? Or even worse duplicate code being tested inside the test itself.

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+using System.Buffers;
+using System.Collections.Sequences;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public readonly partial struct ReadOnlyBuffer 
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static bool TryGetBuffer(Position begin, Position end, out ReadOnlyMemory<byte> data, out Position next)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: internal

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                    return true;
+            }
+
+            ThrowHelper.ThrowNotSupportedException();
+            next = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static Position Seek(Position begin, Position end, long bytes, bool checkEndReachable = true)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: internal

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            else
+            {
+                cursor = SeekMultiSegment(begin, end, bytes, checkEndReachable);
+            }
+
+            return cursor;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static Position SeekMultiSegment(Position begin, Position end, long bytes, bool checkEndReachable)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: private for everything else

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 6 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: Why internal? It can be private

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 8 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: will do

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 7 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: same: why not private?

File: tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs]: What do you propose I do? Is there a way to test this without calling ROB's private APIs?

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 7 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: So the tests from tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs  do not require copied code or being moved.

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 6 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: The same thing, there is a test in tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs  using it.

File: tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs (hunk 2 above)
Comment (pakrym) [reply tests/System.Buffers.Primitives.Tests/ReadableBufferFacts.cs]: Make them internal. It has been done in other places to verify implementation details.

File: tests/System.IO.Pipelines.Tests/TestBufferFactory.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply tests/System.IO.Pipelines.Tests/TestBufferFactory.cs]: There are two copies. I will remove one of them (i.e. the one in pipelines)

File: src/System.Buffers.Primitives/System/Collections/Position.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            _index = 0;
+        }
+
+        public object Segment => _segment;
+        public int Index => _index;
+
+        public static explicit operator int(Position position) => position._index;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public (T segment, int index) Get<T>()
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Collections/Position.cs]: Do we expect consumers to check if the segment is non-null/default?

File: src/System.Buffers.Primitives/System/Collections/Position.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            ThrowHelper.ThrowInvalidOperationException(ExceptionResource.UnexpectedSegmentType);
+            return default;
+        }
+
+        public static bool operator ==(Position left, Position right) => left._index == right._index && left._segment == right._segment;
+        public static bool operator !=(Position left, Position right) => left._index != right._index || left._segment != right._segment;
+
+        public static Position operator +(Position value, int index)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Collections/Position.cs]: I thought we were going to limit the ability to modify positions?

File: src/System.IO.Pipelines.Extensions/ReadableBufferSequence.cs [hunk 11]
```diff
@@ -16,16 +15,16 @@ public ReadableBufferSequence(ReadOnlyBuffer buffer) : this()
             _buffer = buffer;
         }
 
-        public Collections.Sequences.Position First => Collections.Sequences.Position.Create(_buffer.Start.Segment, _buffer.Start.Index);
+        public Position Start => new Position(_buffer.Start.Segment, _buffer.Start.Index);
```
Comment (pakrym) [inline src/System.IO.Pipelines.Extensions/ReadableBufferSequence.cs]: Why have start at all if we treat `default` as a `Start`?

File: src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Hmmm, I have done it initially and it did not work. But now it does. I will fix.

Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: I removed it and added InternalsVisible from Pipelines.Tests to ReadOnlyBuffer. We need to design proper seek APIs though.

Comment (pakrym) [reply src/System.IO.Pipelines/System/IO/Pipelines/PositionOperations.cs]: Wait, why do you need `InternalsVisisbleTo` to `Pipelines.Tests`? Just don't move `SeekTests.cs` from primitives.

File: src/System.Buffers.Primitives/System/Collections/Position.cs (hunk 10 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Collections/Position.cs]: We now have public ctors, Index, and Segment. The + operator is much more hidden than those. 
Also, without the operator, code that uses position as index is very wacky. It does not care about the Segment, yet it has to deal with it (in the ctor, at least).

We can think about it more, but I would treat this as a separate PR.

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs [hunk 12]
```diff
@@ -10,8 +10,65 @@ namespace System.Buffers
 {
     public readonly partial struct ReadOnlyBuffer 
     {
+        public static int Seek(Position begin, Position end, out Position result, byte byte0)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: Change them to be extension/instance methods on `ReadOnlyBuffer`

File: src/System.Buffers.Primitives/System/Collections/Position.cs (hunk 9 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Collections/Position.cs]: Depends on what they use the position for.

File: src/System.IO.Pipelines.Extensions/ReadableBufferSequence.cs (hunk 11 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines.Extensions/ReadableBufferSequence.cs]: I dont think we treat default as start. I remove this early this week.

File: src/System.Buffers.Primitives/System/Collections/Position.cs (hunk 9 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Collections/Position.cs]: I think in most cases they either use it or not. If they don't use it, direct access to Index is enough. If they need both then I prefer throwing to returning default.

File: src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs (hunk 12 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: Let's do it separately. I think I have seen some code that calls the APIs with positions other that the buffer's start/end

Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer_helpers.cs]: I think `Seek` is the last one, but we can do it in a separate pass, just don't like moving things back and forth.

File: src/System.Buffers.Primitives/System/Collections/Position.cs (hunk 9 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Collections/Position.cs]: Sure, we can play with this. But I still want to do an experiment with changing the check to typeof(T) == typeof(Segment). I will try to do it after the new year. If if does not work, we can remove this method.
